aptos-move/framework/aptos-framework/sources/HashMap.move

```diff
@@ -0,0 +1,282 @@
+module AptosFramework::HashMap {
+    use Std::Errors;
+    use Std::Vector;
+    use Std::Hash::sip_hash;
+    use AptosFramework::Table::{Self, Table};
+
+    const MAX_LOAD_PER_BUCKET: u64 = 10;
+    const SPLIT_THRESHOLD: u64 = 75;
+
+    /// Not found in the table;
+    const ENOT_FOUND: u64 = 0;
+    /// Initial capacity needs to be power of two.
+    const ENOT_POWER_OF_TWO: u64 = 1;
+    /// Destroy non-empty hashmap.
+    const ENOT_EMPTY: u64 = 2;
+
+    /// HashMap entry contains both the key and value.
+    struct Entry<K, V> has store {
+        hash: u64,
+        key: K,
+        value: V,
+    }
+
+    /// A hashmap implementation based on linear hashing. (https://www.cs.cmu.edu/afs/cs.cmu.edu/user/christos/www/courses/826-resources/PAPERS+BOOK/linear-hashing.PDF)
+    /// Compare to Table, it uses less storage slots but has higher chance of collision, it's a trade-off between space and time.
+    /// Compare to other hash map implementation, linear hashing splits one bucket a time instead of doubling buckets when expanding to avoid unexpected gas cost.
+    struct HashMap<K, V> has store {
+        buckets: Table<u64, vector<Entry<K, V>>>,
+        num_buckets: u64,
+        // number of bits to represent num_buckets
+        level: u8,
+        // total number of items
+        len: u64,
+    }
+
+    /// Create an empty HashMap with `initial_buckets` buckets.
+    public fun new<K: drop + store, V: store>(initial_buckets: u64): HashMap<K, V> {
+        assert!(initial_buckets > 0 && (initial_buckets & (initial_buckets - 1) == 0), Errors::invalid_argument(ENOT_POWER_OF_TWO));
```

Comment:
Can you explain why this needs to be a power of two? Also, may be add a comment here.

Reply:
+1. If you really have a precondition like this, it must be well documented.